fix(libsinsp): fix some path handling in fs.path

incertum opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines in the [CONTRIBUTING.md](https://github.com/falcosecurity/.github/blob/master/CONTRIBUTING.md) file and learn how to compile Falco from source [here](https://falco.org/docs/source).
2. Please label this pull request according to what type of issue you are addressing.
3. Please add a release note!
4. If the PR is unfinished while opening it specify a wip in the title before the actual title, for example, "wip: my awesome feature"
-->

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

/kind bug

/kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

> /kind feature

<!--
Please remove the leading whitespace before the `/kind <>` you uncommented.
-->

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area API-version

> /area build

> /area CI

> /area driver-kmod

> /area driver-bpf

> /area driver-modern-bpf

> /area libscap-engine-bpf

> /area libscap-engine-gvisor

> /area libscap-engine-kmod

> /area libscap-engine-modern-bpf

> /area libscap-engine-nodriver

> /area libscap-engine-noop

> /area libscap-engine-source-plugin

> /area libscap-engine-savefile

> /area libscap

> /area libpman

/area libsinsp

> /area tests

> /area proposals

<!--
Please remove the leading whitespace before the `/area <>` you uncommented.
-->

**Does this PR require a change in the driver versions?**

> /version driver-API-version-major

> /version driver-API-version-minor

> /version driver-API-version-patch

> /version driver-SCHEMA-version-major

> /version driver-SCHEMA-version-minor

> /version driver-SCHEMA-version-patch

<!--
Please remove the leading whitespace before the `/version <>` you uncommented.
-->

**What this PR does / why we need it**:

`fs.path*` still had some minor bugs in handling path resolution and ensuring no trailing `/` is in the output fields.

**Which issue(s) this PR fixes**:

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
If PR is `kind/failing-tests` please post the related issues/tests in a comment and do not use `Fixes`.
-->

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

<!--
If no, you have to do nothing.
If yes, a release note is required:
Delete `NONE` and enter your extended release note in the block below.
Please note, the release note follows the "conventional commit specification" (https://www.conventionalcommits.org/en/v1.0.0/):
For example: `fix: broken link`.
If the PR requires additional action from users switching to the new release, prepend the string "action required:".
For example, `action required: change the API interface of libscap`.
-->

```release-note
fix(libsinsp): fix some path handling in fs.path
```

incertum commented:
/milestone 0.14.0

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/1571#" title="Author self-approved">incertum</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [incertum]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

incertum commented on userspace/libsinsp/sinsp_filtercheck_fspath.cpp:
```diff
@@ -349,7 +348,14 @@ uint8_t* sinsp_filter_check_fspath::extract(sinsp_evt* evt, OUT uint32_t* len, b
-		m_tstr = sinsp_utils::concatenate_paths(tinfo->get_cwd(), m_tstr);
+		if(m_tstr.front() != '/')
+		{
+			m_tstr = sinsp_utils::concatenate_paths(tinfo->get_cwd(), m_tstr.c_str());
```
@federico-sysdig and @LucaGuerra hope you are ready for the next round :woman_facepalming:
Similar to the last PR we also have some issues here with the string. The possible trailing slash would not be removed otherwise. I have adjusted the unit tests to test this as well. Plus we also need to call `concatenate_paths` when the path is absolute. Absolute paths with path traversal happen on real-life servers.

incertum commented on userspace/libsinsp/test/events_fspath.ut.cpp:
```diff
@@ -27,17 +27,21 @@ class fspath : public sinsp_with_test_input
 	const char *resolved_rel_filename_nopath = "/root/nopath";
 
-	const char *name = "/tmp/name";
+	const char *name = "/tmp/random/dir...///../../name/";
```
Test for possible trailing `/` and path traversal for absolute paths.

incertum commented on userspace/libsinsp/test/events_fspath.ut.cpp:
```diff
@@ -181,42 +185,10 @@ class fspath : public sinsp_with_test_input
-		verify_fields(evt, expected_name, expected_name, NULL, NULL, NULL, NULL);
-	}
-
-	void test_exit_path_raw(const char *expected_name,
```
We could convert `test_exit_path_raw` to `test_exit_path`. That way the unit tests are clearer and more functional.

incertum reviewed (commented):
To the best of my testing abilities `fd.name` and `fs.path.name` now behave consistently the same way.

federico-sysdig commented on userspace/libsinsp/sinsp_filtercheck_fspath.cpp:
```diff
@@ -349,7 +348,14 @@ uint8_t* sinsp_filter_check_fspath::extract(sinsp_evt* evt, OUT uint32_t* len, b
+		} else
+		{
+			// concatenate_paths takes care of resolving the path
+			m_tstr = sinsp_utils::concatenate_paths("", m_tstr.c_str());
```
`std::string` should be automatically cast to `std::string_view`, I don't think you need the `c_str()` call (which needs an automatic cast from `const char*` to `std::string_view` anyway). A similar simplification can be done a few lines above.

incertum replied:
We had that discussion before, nothing else worked here ...

federico-sysdig replied:
Please, forgive me, I forgot. I'm aging. I'm noticing that we do pass a straight `std::string` to this function elsewhere. Do you mean that in this specific call instance it is not working? As in not compiling or not doing the right concatenation?

incertum replied:
It's not doing the right concatenation. I developed and tested on my fedora Linux box.

federico-sysdig reviewed (commented):
Wrote just a few comments.

incertum commented on userspace/libsinsp/utils.cpp:
```diff
@@ -632,7 +632,7 @@ std::string sinsp_utils::concatenate_paths(std::string_view path1, std::string_v
 	// If the path ends with a separator, remove it, as the OS does.
 	//
-	if (result.length() > 1 && result.back() == '/')
+	if (!result.empty() && result.back() == '/')
```
Can revert later. From my perspective our string handling isn't very transparent (at least to me), when do we expect null terminated etc, in a different place in the other PR we use `.empty()`  in another place. The fact we continue having problems https://github.com/falcosecurity/libs/pull/1571#discussion_r1424954825 means something is off or not consistent throughout our code base. WDYT?

federico-sysdig replied:
I don't understand. My comment was not on the technique to check whether a string is null. I was noticing that testing the length for begin at least one character (`str.length() > 1`) isn't exactly the same as testing if the string is not empty (`!str.empty()`). Changing one into the other makes the function behave differently, is this intentional?

LucaGuerra replied:
I agree with @federico-sysdig :)
However, re. @incertum 's uncertainity about string handling, it's an actual problem: strings should behave like strings and not have extra nulls at the end, see the fix in https://github.com/falcosecurity/libs/pull/1574 . If we find other instances of such things (hopefully there aren't) we need to fix them :)

incertum replied:
Apologies yes reverted.

LucaGuerra replied:
> hopefully there aren't

There are :) I will better explain in the comments

incertum commented:
Rebased and squashed with the solution we all want, but I am now still getting some errors locally:

```
[  PASSED  ] 417 tests.
[  FAILED  ] 4 tests, listed below:
[  FAILED  ] fspath.unlinkat_2
[  FAILED  ] fspath.open
[  FAILED  ] fspath.openat_2
[  FAILED  ] fspath.openat2
```

all similar like:
```
Expected equality of these values:
  get_field_as_string(evt, fs_path_name).c_str()
    Which is: "/root/tmp/name/"
  expected_name
    Which is: "/root/tmp/name"
```

incertum commented:
CI is confirming these failures.

LucaGuerra commented:
@incertum the problem lies in the `m_tstr.assign()` calls in that file. As you can see the are taking a raw buffer (incl. NUL terminator) and converting it into a string by giving it the total length so they will always include the NUL. We don't want that, so we should change those assignments to build the string only up to its length to e.g.:

```c++
m_tstr.assign((const char*) extract_values[0].ptr, strnlen((const char*) extract_values[0].ptr, extract_values[0].len));
```

At this point I think we can safely remove the part that says:

```c++
	// If m_tstr ends in a c-style \0, remove it to be
	// consistent. Generally, the evt.rawarg fields above *will*
	// end in c-style \0 characters, while the ones that extract
	// from the enter event or have values populated by parsers
	// (e.g. fchmod/fchown) will not.
	if(m_tstr.back() == '\0')
	{
		m_tstr.pop_back();
	}
```

This becomes redundant because we already deal with the problem earlier as we should, I think it's good to maintain the invariant that `m_tstr` is a well formed string.

LucaGuerra commented on userspace/libsinsp/sinsp_filtercheck_fspath.cpp:
```diff
@@ -349,7 +348,14 @@ uint8_t* sinsp_filter_check_fspath::extract(sinsp_evt* evt, OUT uint32_t* len, b
 		}
 
-		m_tstr = sinsp_utils::concatenate_paths(tinfo->get_cwd(), m_tstr);
+		if(m_tstr.front() != '/')
```
once the problem explained below is fixed I believe we can safely revert this change.

incertum replied:
Follow up: I think we still need this as in this case we concatenate cwd and m_tstr, and else we still use the helper to possibly resolve the path.

FedeDP replied:
Can we use `std::filesystem::path::is_absolute` instead? Much more readable IMHO.
If i got this correctly, we are stating that if `m_tstr` is relative, we attach the current working directory, else we use it as is, right?

incertum replied:
> else we use it as is, right?

almost correct we still need to resolve the absolute path as absolute paths can still contain path traversal. Statement based on real data observed.

incertum commented:
Re https://github.com/falcosecurity/libs/pull/1571#issuecomment-1854818644 thanks @LucaGuerra :rocket: -- suggestions incorporated plus I learned dos and donts with string handling for the future, thanks for that!

LucaGuerra reviewed (approved):
LGTM

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/1571#" title="Author self-approved">incertum</a>*, *<a href="https://github.com/falcosecurity/libs/pull/1571#pullrequestreview-1782198040" title="Approved">LucaGuerra</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [LucaGuerra,incertum]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

poiana commented:
LGTM label has been added.  <details>Git tree hash: 49f3589de256b0d33ac5ae69763f9ead2be2293d</details>

FedeDP reviewed (approved):
/approve

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/1571#pullrequestreview-1782203175" title="Approved">FedeDP</a>*, *<a href="https://github.com/falcosecurity/libs/pull/1571#" title="Author self-approved">incertum</a>*, *<a href="https://github.com/falcosecurity/libs/pull/1571#pullrequestreview-1782198040" title="Approved">LucaGuerra</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [FedeDP,LucaGuerra,incertum]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->
